specification.md

```diff
@@ -0,0 +1,257 @@
+
+## Introduction
+
+The Makefile Specification defines standard interfaces which MUST or SHOULD be implemented in a cloud native application's Makefile.
+
+## Specification
+
+### Overall
+
+The Options in following sections means env variables which can change default action items of Makefile. 
+
+The severity `REQUIRED` means that you MUST support the option. 
+
+The severity `OPTIONAL` means that you MAY support the option.
+
+### Usage for targets
+
+The Makefile MUST print usage message for make targets when `$HELP == y`
+
+#### Example
+
+```bash
+HELP=y make build
+# Build code.
+# make all == make build
+#
+# Args:
+#   WHAT: Directory names to build.  If any of these directories has a 'main'
+#     package, the build will produce executable files under bin/.
+#     If not specified, "everything" will be built.
+#   GOFLAGS: Extra flags to pass to 'go' when building.
+#   GOLDFLAGS: Extra linking flags passed to 'go' when building.
+#   GOGCFLAGS: Additional go compile flags passed to 'go' when building.
+#
+# Example:
+#   make
+#   make all or make build
+#   make build WHAT=cmd/server GOFLAGS=-v
+#   make all GOGCFLAGS="-N -l"
+#     Note: Use the -N -l options to disable compiler optimizations an inlining.
+#           Using these build options allows you to subsequently use source
+#           debugging tools like delve.
+```
+
+### Color log
+
+The Makefile SHOULD print colorized log when `$COLOR_LOG == true`
+
+### Version
+
+The valid version MUST be able to describe the current status of git tree.
+
+A complete version information contains:
+
+-   a git tag versioned by [Semantic Versioning 2.0.0](http://semver.org/spec/v2.0.0.html) (semver). e.g. v1.0.0 v0.1.0-alpha.1
+-   commits number after latest git tag and latest commit hash
+-   git tree status: 
+    -   `clean` indicates no changes since the git commit id.
+    -   `dirty` indicates source code changes after the git commit id
+    -   `archive` indicates the tree was produced by `git archive`
+
+#### Use Case
+
+| version                 | docker tag              | description                              |
+| ----------------------- | ----------------------- | ---------------------------------------- |
+| v0.0.1                  | v0.0.1                  | a clean git tree with the latest tag `v0.0.1` |
+| v0.0.1-dirty            | v0.0.1-dirty            | a dirty git tree with changes not staged |
+| v0.0.1-2+1b4531e1acf800 | v0.0.1-2-1b4531e1acf800 | a clean git tree with a tag ` v0.0.1`.<br />There are two new commits after the tag and the latest commit is `1b4531e1acf800` |
```

Comment:
kubernetes defines version in this way，I don't know the root cause.
I just redefine the docker tag, replacing `+` with `-`, because `+` is not a valid character in docker tag's format.

It is important that the second and third cases MUST only be used for debugging and testing to alert developer that you are working on a dirty git tree or you need to create a new git tag before release.

For release:
1. the final version MUST follow the Semantic Versioning 2.0.0 (semver). e.g. v1.0.0 v0.1.0-alpha.1
2. the git tree MUST be clean
3. the latest tag MUST be even with the latest commit id